decision_service/dsclientlib/ds_driver.cc

```diff
@@ -0,0 +1,131 @@
+
+
+//this macro assumes that success_code equals 0
+#define TRY_OR_RETURN(x) do { \
+  int retval = (x); \
+  if (retval != 0) { \
+    return retval; \
+  } \
+} while (0)
+
+namespace decision_service {
+
+	int check_null_or_empty(const char * arg1, const char * arg2, api_status* status);
+
+	class driver_impl {
+	public:
+    	using error_fn = void(*)(const api_status&,void* user_context);
+
+	private:
+	  	utility::config_collection _configuration;
+    	error_callback_fn _error_cb;
+    	logger _logger;
+
+	public:
+		int ranking_request(const char * uuid, const char * context, ranking_response & response, api_status * status)
+		{
+	    	//clear previous errors if any
+      		if (status != nullptr) status->clear();
+
+			//check arguments
+			TRY_OR_RETURN(check_null_or_empty(uuid, context, status));
+
+			/* GET ACTIONS PROBABILITIES FROM VW */
+
+			/**/ // TODO: replace with call to parse example and predict
+			/**/ // TODO: once that is complete
+				/**/ std::vector<std::pair<int, float>> action_proba;
+			/**/ std::string model_id = "model_id";
+			/**/ action_proba.push_back(std::pair<int, float>(2, 0.4f));
+			/**/ action_proba.push_back(std::pair<int, float>(1, 0.3f));
+			/**/ action_proba.push_back(std::pair<int, float>(4, 0.2f));
+			/**/ action_proba.push_back(std::pair<int, float>(3, 0.1f));
+
+			//send the ranking event to the backend
+			ranking_event evt(uuid, context, action_proba, model_id);
+			TRY_OR_RETURN(_logger.append_ranking(evt.serialize(), status));
+
+			response.set_uuid(uuid);
+			response.set_ranking(action_proba);
+
+			return error_code::success;
+		}
+
+		//here the uuid is auto-generated
+		int ranking_request(const char * context, ranking_response & response, api_status * status)
+		{
+			return ranking_request(context, boost::uuids::to_string(boost::uuids::random_generator()()).c_str(), response, status);
+		}
+
+		int report_outcome(const char* uuid, const char* outcome_data, api_status * status)
+		{
+	    	//clear previous errors if any
+      		if (status != nullptr) status->clear();
+
+			//check arguments
+			TRY_OR_RETURN(check_null_or_empty(uuid, outcome_data, status));
+
+			//send the outcome event to the backend
+			outcome_event evt(uuid, outcome_data);
+			TRY_OR_RETURN(_logger.append_outcome(evt.serialize(), status));
+
+			return error_code::success;
+		}
+
+		int report_outcome(const char* uuid, float reward, api_status * status)
+		{
+			return report_outcome(uuid, std::to_string(reward).c_str(), status);
+		}
+
+		driver_impl(const utility::config_collection& config, error_fn fn = nullptr, void* err_context = nullptr)
+			: _configuration(config),
+        _error_cb(fn,err_context),
+  			_logger(config,&_error_cb)
+		{
+		}
+
+	};
+
+	//driver implementation
+	driver::~driver() {}
+
+	driver::driver(const utility::config_collection& config, error_fn fn, void* err_context) 
+	{
+		_pimpl = std::unique_ptr<driver_impl>(new driver_impl(config,fn,err_context));
```

Comment:
>std::unique_ptr<driver_impl> [](start = 11, length = 28)

can't we do simple pointers and deallocate in the destructor. that way we don't need to expose std::unique_ptr. Or do you gain anything else from unique_ptr?

Reply:
Makes ownership and lifetime of the pointer clear.  The de-allocation is done automatically.  However unique_ptr does show up in the header file.  Removed it since its a simple change.

---
In reply to: [186736006](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186736006) [](ancestors = 186736006)